### PR TITLE
Benja press mention rework 901 905

### LINKED ISSUE #901: Fetch and use more OG data in presse mentions
I think we should fetch more OG data (and use it, unless the user overrides).  And we should do this as soon as we have the URl, not encourage the user to fill out a form that we're going to auto-fill and ignore in a moment, after they hit "save".

So, a new pm asks you for URL (only).  Once you provide it and confirm, it does the OG fetch, makes things like journal name, article title, and article description be copies of what it found in OG, saves that, and displays the pm as a view-one.  Click to edit.

There's an open question whether we should go to view-one or edit-one at that step two.  It will depend on experience, on whether it's rare or common that the OG data doesn't suffice.

It's especially frustrating that journal title, article title, and article description or mandatory, and yet all three can generally be fetched from OG data.

### PR DESCRIPTION
**Pull request by Shriukan33**
this PR : 

* Adds Unique constraint to PressMention's URLs, so no URL can be saved twice.
* Drops copyright warning that was too important on screen while irrelevant
* Add a pre-creation form that will only require the Article's URL you want to submit, and fetch OG data in it to pre-fill the creation form.
* Creation form's content now trumps OG data, so what the user inputs is showed, while OG is now suggesting content automatically.

Closes #901 #902 #905

### CODE REVIEW

**Comment by JeffAbrahamson** [comment]
Why is there a hard minimum on this field?

image

**Comment by JeffAbrahamson** [comment]
This is really confusing.  I changed the URL.  If I can do that, that means you're ready to change your processing.  And, indeed, you fetched some new og data but you're still showing the original article title and description, fetched from og, but I didn't modify them.

If the user changes the URL, there needs to be feedback about what happened.  It has to be transparent for the user.  For example, you'll recall we discussed making this a separate back-and-forth like the way github lets you change issue name?

image

**Comment by JeffAbrahamson** [comment]
You're not fetching or storing og:site_name.

**Comment by JeffAbrahamson** [comment]
* You're not creating a new publication, you're creating a new citation.  Or something like that.
* The journal name is missing (og_site_name).
* The publication date is today's date.  If you don't have something like `og:article:published_time`, then you don't know the date and you shouldn't provide it for the user.

A reminder, open graph is a protocol, it has a [specification](https://opengraphprotocol.org/).

image

**Comment by JeffAbrahamson** [comment]
I have yet to find an article under 200 characters.

image

**Comment by JeffAbrahamson** [comment]
This should open in a new tab.  We do not want to take people away from our site.

http://localhost:8000/presse/admin/detail/8

**Comment by JeffAbrahamson** [comment]
* These should also open in new tabs.
* The journal title is wrong, it is the base of the URL (which may or may not be meaningful), whereas you stored `newspaper_name` in the db.
* The images should be clickable, that is _the_ hotspot for clicks and it is dead here.
* Only four articles at a time on a big screen?  Is that reasoned?
* Where are the article descriptions (possibly truncated to N words with ellipsis)?

image

**Review by JeffAbrahamson — CHANGES_REQUESTED** [review]

**Comment by Shriukan33** [comment]
> This is really confusing. I changed the URL. If I can do that, that means you're ready to change your processing. And, indeed, you fetched some new og data but you're still showing the original article title and description, fetched from og, but I didn't modify them.
> 
> If the user changes the URL, there needs to be feedback about what happened. It has to be transparent for the user. For example, you'll recall we discussed making this a separate back-and-forth like the way github lets you change issue name?
> 
> Why does it say "www.lemonde.fr", that's neither the name of the journal nor stored in the db?
> 
> image

I'm confused about this one, I don't understand what you went through, what you did, what you expected, what happened ?

**Comment by JeffAbrahamson** [comment]
> > This is really confusing. I changed the URL. If I can do that, that means you're ready to change your processing. And, indeed, you fetched some new og data but you're still showing the original article title and description, fetched from og, but I didn't modify them.
> > If the user changes the URL, there needs to be feedback about what happened. It has to be transparent for the user. For example, you'll recall we discussed making this a separate back-and-forth like the way github lets you change issue name?
> > Why does it say "www.lemonde.fr", that's neither the name of the journal nor stored in the db?
> > image
> 
> I'm confused about this one, I don't understand what you went through, what you did, what you expected, what happened ?

The article is from PO, the photo is from LM, the journal title is a hostname and not a journal title.

**Comment by Shriukan33** [comment]
> The article is from PO, the photo is from LM, the journal title is a hostname and not a journal title.

So you originally entered the LM article, saved, then changed the quotation, and then changed what ? The URL only ?

**Comment by JeffAbrahamson** [comment]
> > The article is from PO, the photo is from LM, the journal title is a hostname and not a journal title.
> 
> So you originally entered the LM article, saved, then changed the quotation, and then changed what ? The URL only ?

Did you try it?

I didn't even save.  I entered the PO URL, it did the fetch.  I changed the URL to LM, then saved.

**Comment by Shriukan33** [comment]
> > > The article is from PO, the photo is from LM, the journal title is a hostname and not a journal title.
> > 
> > 
> > So you originally entered the LM article, saved, then changed the quotation, and then changed what ? The URL only ?
> 
> Did you try it?
> 
> I didn't even save. I entered the PO URL, it did the fetch. I changed the URL to LM, then saved.

the fetch happens when you enter the url in the pre-form, to pre-fill the fields. Any change past that point is user input and trumps whatever OG was saying.
So yeah, if you fetch OG image and then change the URL, it keeps the old one because the fetch happened earlier. Why would you do a fetch on a url, have its og data prefill, only to change everything after ? It doesn't make sense ?

**Comment by JeffAbrahamson** [comment]
> > I didn't even save. I entered the PO URL, it did the fetch. I changed the URL to LM, then saved.
> 
> the fetch happens when you enter the url in the pre-form, to pre-fill the fields. Any change past that point is user input and trumps whatever OG was saying. So yeah, if you fetch OG image and then change the URL, it keeps the old one because the fetch happened earlier. Why would you do a fetch on a url, have its og data prefill, only to change everything after ? It doesn't make sense ?

Users aren't obliged to make sense to developers who are sitting in other places at other times. ;-)  In testing, I'm trying to do things that I could imagine a user doing.  Like pasting the wrong URL and only realising it when the OG displays.

I sketched out a possible workflow response path in #901.

**Comment by JeffAbrahamson** [comment]
ac27cfc4ee6 and ec1c49a LGTM, let's at least merge those to reduce PR complexity.

**Comment by Shriukan33** [comment]
Updating this PR : 
* Added a button in creation form to refresh OG data after initial POST, for clumsy users :innocent: 
* Improved responsive on list view, you should be able to see more articles per page on large screen
* Added the description in the list (previously using incorrect attribute)
* Reworded bits of the text you mentioned to be less verbose 
* Increased text limits to 1000 so we're not bothered anymore
* Added og:site_name prefill
* Articles now open in new tabs
* Images are now clickable

**Comment by Shriukan33** [comment]
> [ac27cfc](https://github.com/transport-nantes/tn_web/commit/ac27cfc4ee691824872caa08e9f214f7d4160e8e) and [ec1c49a](https://github.com/transport-nantes/tn_web/commit/ec1c49a2280a629dfcfb0cd9c00ff3117b7fba50) LGTM, let's at least merge those to reduce PR complexity.

Done

**Comment by Shriukan33** [comment]
update :

* Changing URL in creation form now updates the associated opengraph data
* Fixed a minor discrepancy between detailview and listview

File: transport_nantes/press/views.py [hunk 1]
```diff
@@ -96,21 +101,99 @@
...
+    - a tuple containing the title, description, image, newspaper_name and
+    publication_date if the function is called from a view
+    - a JSONResponse containing the title, description, image, newspaper_name
+    and publication_date if the function is called from browser (Ajax)
+    """
+    if not isView:
+        url = request.POST["url"]
+
+    try:
+        response = requests.get(url)
```
[the remark this answers is not in the document]
Comment (Shriukan33) [reply transport_nantes/press/views.py:146]: This bot is right, it leaves us open to get on any user supplied url.
We would curate the users that have that permission, but still ... I don't see how we could add more security layers.

[the remark this answers is not in the document]
Comment (JeffAbrahamson) [reply transport_nantes/press/views.py:146]: That's why such an ajax endpoint should be POST-only and should 403 on GET.  That permits you to use a CSRF token to validate the request.

We don't want to be an OG fetching server, indeed.

[the remark this answers is not in the document]
Comment (JeffAbrahamson) [reply transport_nantes/press/views.py:146]: That said, this page is not available except to people with the appropriate permission.  (You did add permissions to pressmention, iirc, right?)  At some point, we are obliged to respond to user input on a website.

[the remark this answers is not in the document]
Comment (JeffAbrahamson) [reply transport_nantes/press/views.py:146]: Ah, and it's noting SSRF attacks.  You'll have to google it, but there are ways to validate that it's a valid URL and doesn't include stuff that will lead to further parsing down the line and further action.

**Comment by Shriukan33** [comment]
PR is moved to #910

**Pull request closed without merging**